application/src/main/java/org/thingsboard/server/actors/device/DeviceActorMessageProcessor.java

```diff
@@ -705,10 +706,19 @@ private void processRpcResponseStatus(SessionInfoProto sessionInfo, ToDeviceRpcR
                 maxRpcRetries = maxRpcRetries == null ?
                         systemContext.getMaxRpcRetries() : Math.min(maxRpcRetries, systemContext.getMaxRpcRetries());
                 if (maxRpcRetries <= md.getRetries()) {
-                    toDeviceRpcPendingMap.remove(requestId);
-                    status = RpcStatus.FAILED;
-                    response = JacksonUtil.newObjectNode().put("error", "There was a Timeout and all retry " +
-                            "attempts have been exhausted. Retry attempts set: " + maxRpcRetries);
+                    if (closeTransportSessionOnRpcDeliveryTimeout) {
+                        md.setRetries(0);
+                        status = RpcStatus.QUEUED;
```

Comment:
Why do we use QUEUED status in this case?

Reply:
Customer request.

> Customer asked to add a configuration parameter that will allow closing the session immediately in case of RPC delivery timeout and return RPC to the QUEUED state, so they can decide on their own if the RPC needs to be removed from the map of pending RPCs based on RPC status updates history.

Reply:
Just to clarify about RPC status updates history: 

With a default TB functionality, we can't see a history of RPC updates. However, since we are sending rule engine notifications about RPC status changes, we can track the history of RPC updates based on those events. This is basically how we do this.